ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+logs being shown in the command-line _(9)_. The developer can interactively watch the container build
+execution and possible errors that might surface.
+
+At the end, when the build-pod is completed successfully _(5)_, the CLI shows the meta information
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][ClientGoWatch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
```

Comment:
Hm, you define this on the BuildRun. I think the same should be definable on the Build. If we hereby introduce overriding the Build's sources, then maybe more details are necessary to describe how the behavior should be.

In general I am not sure if this is necessary in this ship. I guess that somebody who runs a build from local, will use a dedicated Build anyway because that user will also want to push to a different location.

Eventually, we should still have that feature, though.

Reply:
> Hm, you define this on the BuildRun. I think the same should be definable on the Build. If we hereby introduce overriding the Build's sources, then maybe more details are necessary to describe how the behavior should be.
>
> In general I am not sure if this is necessary in this ship. I guess that somebody who runs a build from local, will use a dedicated Build anyway because that user will also want to push to a different location.
> 
> Eventually, we should still have that feature, though.

My arguments for it are on this [previous comment](https://github.com/shipwright-io/community/pull/11#discussion_r684971517). Please note I'm suggesting we use a simple annotation initially, but when the multiple sources EP comes along we already have a suggestion of how to make those two features work together.